Prevent non admin from changing admin attributes

cbliard opened the pull request:
https://community.openproject.org/wp/50121

cbliard committed:
Fix multiline #describe text in specs

cbliard committed:
refactor spec: differentiate user being requested and user requesting

cbliard committed:
refactor: remove useless return

Making rubocop happier every time.

cbliard committed:
Add controller test for user admin flag update

cbliard committed:
Forbid any update on admin unless user is admin themselves

This means a user with manage_user global permission can no longer update admin attributes.

cbliard committed:
Ensure admins cannot be edited by anybody but active admins

cbliard committed:
Hide user edit form tab if not allowed to edit their attributes

Only admins can edit another admin attributes

cbliard committed:
Update documentation about edit users global permission

aaron-contreras commented on docs/system-admin-guide/users-permissions/roles-permissions/README.md:
```diff
@@ -125,11 +125,15 @@ To delete an existing role click on the **delete icon** next to a role in the li
 
+### Can a user with "Edit users" global permission change administrators attributes?
+
+No, only Administrators can update other Administrators attributes like name or email. This is to prevent the possibility of a user with "Edit users" global permission to impersonate an Administrator by changing the email address to match theirs.
```
```suggestion
No, only Administrators can update other Administrators attributes like name or email. This is to prevent the possibility of a user with "Edit users" global permission impersonating an Administrator by changing the email address to match theirs.
```

aaron-contreras commented on spec/features/users/edit_users_spec.rb:
```diff
@@ -84,13 +84,28 @@ def user_password
     end
   end
 
+  def have_tab_visible(label)
```
```suggestion
  def have_visible_tab(label)
```

aaron-contreras commented on spec/features/users/edit_users_spec.rb:
```diff
@@ -84,13 +84,28 @@ def user_password
+      another_admin = create(:admin)
+      visit edit_user_path(another_admin)
+
+      expect(page).to have_tab_visible('GENERAL')
```
```suggestion
      expect(page).to have_visible_tab('GENERAL')
```

aaron-contreras commented on spec/features/users/edit_users_spec.rb:
```diff
@@ -84,13 +84,28 @@ def user_password
     it 'can too edit the user' do
       visit edit_user_path(user)
 
+      expect(page).to have_tab_visible('GENERAL')
```
```suggestion
      expect(page).to have_visible_tab('GENERAL')
```

aaron-contreras commented on spec/features/users/edit_users_spec.rb:
```diff
@@ -123,9 +138,16 @@ def user_password
+    it 'can not edit attributes of an admin user' do
+      visit edit_user_path(admin)
+
+      expect(page).to have_tab_visible('PROJECTS')
+      expect(page).not_to have_tab_visible('GENERAL')
```
```suggestion
      expect(page).to have_visible_tab('PROJECTS')
      expect(page).not_to have_visible_tab('GENERAL')
```

aaron-contreras commented on spec/features/users/edit_users_spec.rb:
```diff
@@ -123,9 +138,16 @@ def user_password
       visit edit_user_path(user)
 
-      click_on 'Send invitation'
+      click_button 'Send invitation'
```
🖌️

cbliard replied:
Making rubocop happier every day!

aaron-contreras commented on .rubocop.yml:
```diff
@@ -239,6 +239,7 @@ RSpec/ContextWording:
   Prefixes:
     - and
     - as
+    - even
```
I like it but it looks like it wasn't used anywhere?

cbliard replied:
It's used here: https://github.com/opf/openproject/pull/13779/files#diff-5f44c1c580dbc94788e40f1aecf6f711215f30ce917214dc8d25765105cc4a37R306 (you have to click "load diff")

aaron-contreras replied:
Whoops!, thank you

cbliard committed:
Fix broken tests

Use `current_user` instead of `User.current` for the reasons explained in `app/controllers/concerns/accounts/current_user.rb`

cbliard committed:
Update docs/system-admin-guide/users-permissions/roles-permissions/README.md

Co-authored-by: Aaron Contreras <61627014+aaron-contreras@users.noreply.github.com>

cbliard committed:
refactor: Use better English

Co-authored-by: Aaron Contreras <aaronlcaq@gmail.com>

cbliard committed:
Make rubocop happier
